File: 2016/Day-05-How-About-a-Nice-Game-of-Chess/py-2016-05/main_part_2.py
### Source : https://adventofcode.com/2016/day/5#part2
### Author : liuyubobobo
### Time   : 2020-12-04

import hashlib

if __name__ == '__main__':

    # initializing string
    base = "abbhdwsy"

    res = [' ' for _ in range(8)]
    index = 0
    p = 0
    while True:
        str2hash = base + str(index)
        # then sending to md5()
        result = hashlib.md5(str2hash.encode()).hexdigest()

        if result[0:5] == "00000" and '0' <= result[5] <= '7' and res[int(result[5])] == ' ':
            res[int(result[5])] = result[6]

            print(str(p) + " get!")
            p += 1
            if p == 8:
                break

        index += 1

    print(res)

File: 2019/Day-04-Secure-Container/cpp-2019-04/main_part_2.cpp
/// Source : https://adventofcode.com/2019/day/4
/// Author : liuyubobobo
/// Time   : 2020-12-03

#include <iostream>
#include <fstream>
#include <vector>
#include <cassert>
#include <set>
#include <unordered_map>

using namespace std;


class Solution{

public:
    int solve(int a, int b){

        int res = 0;
        for(int i = a; i <= b; i ++)
            res += ok(i);
        return res;
    }

private:
    bool ok(int x){

        string s = to_string(x);
        bool ok = false;
        for(int i = 1; i < s.size(); i ++){
            if(s[i - 1] > s[i]) return false;
            if(s[i - 1] == s[i]){
                if(i - 2 >= 0 && s[i - 2] == s[i]) continue;
                if(i + 1 < s.size() && s[i + 1] == s[i]) continue;
                ok = true;
            }
        }
        return ok;
    }
};

int main() {

    cout << Solution().solve(240920, 789857) << endl;

    return 0;
}
